src/mongo/db/pipeline/accumulator_percentile.cpp

```diff
@@ -0,0 +1,186 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * As a special exception, the copyright holders give permission to link the
+ * code of portions of this program with the OpenSSL library under certain
+ * conditions as described in each individual source file and distribute
+ * linked combinations including the program with the OpenSSL library. You
+ * must comply with the GNU Affero General Public License in all respects for
+ * all of the code used other than as permitted herein. If you modify file(s)
+ * with this exception, you may extend this exception to your version of the
+ * file(s), but you are not obligated to do so. If you do not wish to do so,
+ * delete this exception statement from your version. If you delete this
+ * exception statement from all source files in the program, then also delete
+ * it in the license file.
+ */
+
+#include "mongo/platform/basic.h"
+
+#include "mongo/db/pipeline/accumulator.h"
+
+#include "mongo/db/pipeline/accumulation_statement.h"
+#include "mongo/db/pipeline/document.h"
+#include "mongo/db/pipeline/expression.h"
+#include "mongo/db/pipeline/expression_context.h"
+#include "mongo/db/pipeline/value.h"
+#include "mongo/platform/decimal128.h"
+
+#include "mongo/db/pipeline/TDigest.h"
+
+namespace mongo {
+
+using boost::intrusive_ptr;
+
+REGISTER_ACCUMULATOR(percentile, AccumulatorPercentile::create);
+REGISTER_EXPRESSION(percentile, ExpressionFromAccumulator<AccumulatorPercentile>::parse);
+
+
+const char* AccumulatorPercentile::getOpName() const {
+    return "$percentile";
+}
+
+namespace {
+const char subTotalName[] = "subTotal";
+const char subTotalErrorName[] = "subTotalError";  // Used for extra precision
+const char countName[] = "count";
+}  // namespace
+
+void AccumulatorPercentile::processInternal(const Value& input, bool merging) {
+
+    // Determining 'digest_size'
+    if (this->digest_size == 0){
+        
+        if (input.getDocument()["digest_size"].missing()){
+            this->digest_size = 1000;
+            }
+        else
+            {
+            this->digest_size = input.getDocument()["digest_size"].getDouble();
+            }
+    }
+
+    // ToDo: error codes are not accurate. Set better numbers later
+    // ToDo: It might be better evaluations for this part.
+    uassert(6677, "The 'perc' should be present in the input document.",
+    !input.getDocument()["perc"].missing());
+
+    uassert(6678, "The 'value' should be present in the input document.",
+    !input.getDocument()["value"].missing());
+
+    this->perc_val = input.getDocument()["perc"].getDouble() / 100;  // Converting Percentile to Quantile - [0:100] to [0:1]
```

Comment:
Don't like the naming of 'perc_val'.
You either store the percentile in it (not the quantile).
Or you rename it to some name which makes it obvious that the quantile is store.

Right now the 'quantile' is stored in a field called 'percentile' :(

Reply:
Fixed. Now percentile will be used everywhere with values between 0 and 100